File: iOS/Kalorie/Core/UseCases/FetchFoodItemReportsUseCase.swift
//
//  FetchFoodItemReportsUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 17.09.2026.
//

import Foundation

protocol FetchFoodItemReportsUseCaseProtocol {
    func callAsFunction() async throws -> [FoodItemReportDomain]
}

struct FetchFoodItemReportsUseCase: FetchFoodItemReportsUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol, authProvider: any AuthProviderProtocol) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    func callAsFunction() async throws -> [FoodItemReportDomain] {
        guard authProvider.userId != nil else { throw AuthError.notAuthenticated }
        let dtos: [FoodItemReportDTO] = try await dataProvider.loadAsync(
            from: Constants.Firestore.foodItemReports,
            orderBy: "reported_at",
            descending: true,
            limit: Constants.Firestore.reportsPageLimit
        )
        return dtos.map { $0.asDomain() }
    }
}

#if DEBUG
struct FetchFoodItemReportsUseCaseFake: FetchFoodItemReportsUseCaseProtocol {

    // MARK: - Properties

    var stubbedReports: [FoodItemReportDomain] = []
    var shouldThrow = false

    // MARK: - Functions

    func callAsFunction() async throws -> [FoodItemReportDomain] {
        if shouldThrow { throw URLError(.unknown) }
        return stubbedReports
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/FetchFoodItemReportsUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.models.FoodItemReportDomain
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.networking.FoodItemReportDTO
import antoni.kalorie.core.networking.loadAsync
import antoni.kalorie.core.utils.Constants

interface FetchFoodItemReportsUseCaseProtocol {
    suspend operator fun invoke(): List<FoodItemReportDomain>
}

class FetchFoodItemReportsUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
) : FetchFoodItemReportsUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(): List<FoodItemReportDomain> {
        if (authProvider.userId == null) throw AuthError.NotAuthenticated
        val dtos: List<FoodItemReportDTO> = dataProvider.loadAsync(
            from = Constants.Firestore.FOOD_ITEM_REPORTS,
            orderBy = "reported_at",
            descending = true,
            limit = Constants.Firestore.REPORTS_PAGE_LIMIT,
        )
        return dtos.map { it.asDomain() }
    }
}
